usr/man/man/usr/share/man/man2/lseek.2
'\" te
.\" Copyright 1989, AT&.T. Copyright (c) 2005, Sun Microsystems, Inc.  All Rights Reserved.
.TH lseek 2 "4 May 2005" "SunOS 5.11" "System Calls"
.SH NAME
lseek \- move read/write file pointer
.SH SYNOPSIS
.LP
.nf
#include <sys/types.h>
#include <unistd.h>

\fBoff_t\fR \fBlseek\fR(\fBint\fR \fIfildes\fR, \fBoff_t\fR \fIoffset\fR, \fBint\fR \fIwhence\fR);
.fi

.SH DESCRIPTION
.sp
.LP
The \fBlseek()\fR function sets the file pointer associated with the open file descriptor specified by \fIfildes\fR as follows:
.RS +4
.TP
.ie t \(bu
.el o
If \fIwhence\fR is \fBSEEK_SET\fR, the pointer is set to \fIoffset\fR bytes.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If \fIwhence\fR is \fBSEEK_CUR\fR, the pointer is set to its current location plus \fIoffset\fR.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If \fIwhence\fR is \fBSEEK_END\fR, the pointer is set to the size of the file plus \fIoffset\fR.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If \fIwhence\fR is \fBSEEK_HOLE\fR, the offset of the start of the next hole greater than or equal to the supplied offset is returned. The definition of a hole is provided near the end of the DESCRIPTION.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If \fIwhence\fR is \fBSEEK_DATA\fR, the file pointer is set to the start of the next non-hole file region greater than or equal to the supplied offset.
.RE
.sp
.LP
The symbolic constants \fBSEEK_SET\fR, \fBSEEK_CUR\fR, \fBSEEK_END\fR, \fBSEEK_HOLE\fR, and \fBSEEK_DATA\fR are defined in the header <\fBunistd.h\fR>.
.sp
.LP
Some devices are incapable of seeking. The value of the file pointer associated with such a device is undefined.
.sp
.LP
The \fBlseek()\fR function allows the file pointer to be set beyond the existing data in the file. If data are later written at this point, subsequent reads in the gap between the previous end of data and the newly written data will return bytes of value 0 until data are written into the gap.
.sp
.LP
If \fIfildes\fR is a remote file descriptor and \fIoffset\fR is negative, \fBlseek()\fR returns the file pointer  even if it is negative. The \fBlseek()\fR function will not, by itself, extend the size of a file.
.sp
.LP
If \fIfildes\fR refers to a shared memory object, \fBlseek()\fR behaves as if \fIfildes\fR referred to a regular file.
.sp
.LP
A "hole" is defined as a contiguous range of bytes in a file, all having the value of zero, but not all zeros in a file are guaranteed to be represented as holes returned with \fBSEEK_HOLE\fR. Filesystems are allowed to expose ranges of zeros with \fBSEEK_HOLE\fR, but not required to. Applications can use \fBSEEK_HOLE\fR to optimise their behavior for ranges of zeros, but must not depend on it to find all such ranges in a file. The existence of a hole at the end of every data region allows for easy programming and implies that a virtual hole exists at the end of the file. Applications should use \fBfpathconf\fR(\fB_PC_MIN_HOLE_SIZE\fR) or \fBpathconf\fR(\fB_PC_MIN_HOLE_SIZE\fR) to determine if a filesystem supports \fBSEEK_HOLE\fR. See \fBfpathconf\fR(2).
.sp
.LP
For filesystems that do not supply information about holes, the file will be represented as one entire data region.
.SH RETURN VALUES
.sp
.LP
Upon successful completion, the resulting offset, as measured in bytes from the beginning of the file, is returned. Otherwise, \fB(off_t)\(mi1\fR is returned, the file offset remains unchanged, and \fBerrno\fR is set to indicate the error.
.SH ERRORS
.sp
.LP
The \fBlseek()\fR function will fail if:
.sp
.ne 2
.mk
.na
\fB\fBEBADF\fR\fR
.ad
.RS 13n
.rt  
The \fIfildes\fR argument is not an open file descriptor.
.RE

.sp
.ne 2
.mk
.na
\fB\fBEINVAL\fR\fR
.ad
.RS 13n
.rt  
The \fIwhence\fR argument is not \fBSEEK_SET\fR, \fBSEEK_CUR\fR, or \fBSEEK_END\fR; or the \fIfildes\fR argument is not a remote file descriptor and the resulting file pointer would be negative.
.RE

.sp
.ne 2
.mk
.na
\fB\fBENXIO\fR\fR
.ad
.RS 13n
.rt  
For \fBSEEK_DATA\fR, there are no more data regions past the supplied offset. For \fBSEEK_HOLE\fR, there are no more holes past the supplied offset.
.RE

.sp
.ne 2
.mk
.na
\fB\fBEOVERFLOW\fR\fR
.ad
.RS 13n
.rt  
The resulting file offset would be a value which cannot be represented correctly in an object of type \fBoff_t\fR for regular files.
.RE

.sp
.ne 2
.mk
.na
\fB\fBESPIPE\fR\fR
.ad
.RS 13n
.rt  
The \fIfildes\fR argument is associated with a pipe, a FIFO, or a socket.
.RE

.SH USAGE
.sp
.LP
The \fBlseek()\fR function has a transitional interface for 64-bit file offsets.  See \fBlf64\fR(5).
.sp
.LP
In multithreaded applications, using \fBlseek()\fR in conjunction with a \fBread\fR(2) or \fBwrite\fR(2) call on a file descriptor shared by more than one thread is not an atomic operation.  To ensure atomicity, use \fBpread()\fR or \fBpwrite()\fR.
.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Interface StabilityCommitted
_
MT-LevelAsync-Signal-Safe
_
StandardSee \fBstandards\fR(5).
.TE

.SH SEE ALSO
.sp
.LP
\fBcreat\fR(2), \fBdup\fR(2), \fBfcntl\fR(2), \fBfpathconf\fR(2), \fBopen\fR(2), \fBread\fR(2), \fBwrite\fR(2), \fBattributes\fR(5), \fBlf64\fR(5), \fBstandards\fR(5)


usr/src/uts/common/syscall/lseek.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright (c) 1994, 2011, Oracle and/or its affiliates. All rights reserved.
 */

/*	Copyright (c) 1983, 1984, 1985, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*
 * Portions of this source code were derived from Berkeley 4.3 BSD
 * under license from the Regents of the University of California.
 */

#include <sys/param.h>
#include <sys/isa_defs.h>
#include <sys/types.h>
#include <sys/sysmacros.h>
#include <sys/cred.h>
#include <sys/systm.h>
#include <sys/errno.h>
#include <sys/vnode.h>
#include <sys/file.h>
#include <sys/debug.h>
#include <sys/cmn_err.h>
#include <sys/filio.h>

/*
 * These are defined in unistd.h - but we can't include that
 */
#define	SEEK_SET	0	/* Set file pointer to "offset" */
#define	SEEK_CUR	1	/* Set file pointer to current plus "offset" */
#define	SEEK_END	2	/* Set file pointer to EOF plus "offset" */
#define	SEEK_DATA	3	/* Set file pointer to next data past offset */
#define	SEEK_HOLE	4	/* Set file pointer to next hole past offset */

/*
 * Seek on a file
 */

#if defined(_SYSCALL32_IMPL) || defined(_ILP32)
/*
 * Workhorse for the 32-bit seek variants: lseek32 and llseek32
 *
 * 'max' represents the maximum possible representation of offset
 * in the data type corresponding to lseek and llseek. It is
 * MAXOFF32_T for off32_t and MAXOFFSET_T for off64_t.
 * We return EOVERFLOW if we cannot represent the resulting offset
 * in the data type.
 * We provide support for character devices to be seeked beyond MAXOFF32_T
 * by lseek. To maintain compatibility in such cases lseek passes
 * the arguments carefully to lseek_common when file is not regular.
 * (/dev/kmem is a good example of a > 2Gbyte seek!)
 */
static int
lseek32_common(file_t *fp, int stype, offset_t off, offset_t max,
    offset_t *retoff)
{
	vnode_t *vp;
	struct vattr vattr;
	int error;
	u_offset_t noff;
	offset_t curoff, newoff;
	int reg;

	vp = fp->f_vnode;
	reg = (vp->v_type == VREG);
	if (reg) {
		/*
		 * This is a POSIX requirement for atomicity
		 * of calls to read[v](), write[v]() and lseek()
		 * when operating on regular files.
		 */
		mutex_enter(&fp->f_rwlock);
	}
	curoff = fp->f_offset;

	switch (stype) {
	case SEEK_SET:
		noff = (u_offset_t)off;
		if (reg && noff > max) {
			error = EINVAL;
			goto out;
		}
		break;

	case SEEK_CUR:
		if (reg && off == 0) {	/* regular file, query only */
			error = 0;
			(*retoff) = curoff;
			goto out;
		}
		if (reg && off > (max - curoff)) {
			error = EOVERFLOW;
			goto out;
		}
		noff = (u_offset_t)(off + curoff);
		if (reg && noff > max) {
			error = EINVAL;
			goto out;
		}
		break;

	case SEEK_END:
		vattr.va_mask = AT_SIZE;
		if (error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL)) {
			goto out;
		}
		if (reg && (off  > (max - (offset_t)vattr.va_size))) {
			error = EOVERFLOW;
			goto out;
		}
		noff = (u_offset_t)(off + (offset_t)vattr.va_size);
		if (reg && noff > max) {
			error = EINVAL;
			goto out;
		}
		break;

	case SEEK_DATA:
		/*
		 * Get and set the file pointer to the offset of the next
		 * data past "off"
		 */
		noff = (u_offset_t)off;
		error = VOP_IOCTL(vp, _FIO_SEEK_DATA, (intptr_t)(&noff),
		    FKIOCTL, kcred, NULL, NULL);
		if (error) {
			if (error != ENOTTY)
				goto out;
			/*
			 * The ioctl is not supported, check the supplied
			 * "off" is not past the end of file
			 */
			vattr.va_mask = AT_SIZE;
			error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL);
			if (error)
				goto out;
			if (noff >= (u_offset_t)vattr.va_size) {
				error = ENXIO;
				goto out;
			}
		}
		if (reg && (noff > max)) {
			error = EOVERFLOW;
			goto out;
		}

		fp->f_offset = (offset_t)noff;
		(*retoff) = (offset_t)noff;
		goto out;

	case SEEK_HOLE:
		/*
		 * Get and set the file pointer to the offset of the next
		 * hole past "off"
		 */
		noff = (u_offset_t)off;
		error = VOP_IOCTL(vp, _FIO_SEEK_HOLE, (intptr_t)(&noff),
		    FKIOCTL, kcred, NULL, NULL);
		if (error) {
			if (error != ENOTTY)
				goto out;
			/*
			 * ioctl is not supported, if the off is valid return
			 * the "virtual hole" at the end of the file.
			 */
			vattr.va_mask = AT_SIZE;
			error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL);
			if (error)
				goto out;
			if (off < (offset_t)vattr.va_size)
				noff = (u_offset_t)vattr.va_size;
			else {
				error = ENXIO;
				goto  out;
			}
		}
		if (reg && (noff > max)) {
			error = EOVERFLOW;
			goto out;
		}

		fp->f_offset = (offset_t)noff;
		(*retoff) = (offset_t)noff;
		goto out;

	default:
		error = EINVAL;
		goto out;
	}

	ASSERT((reg && noff <= max) || !reg);
	newoff = (offset_t)noff;
	if ((error = VOP_SEEK(vp, curoff, &newoff, NULL)) == 0) {
		/*
		 * lseek(fd, 0, SEEK_CUR) should not reset the file offset
		 * even if this is not a regular file.
		 */
		if (off != 0 || stype != SEEK_CUR)
			fp->f_offset = newoff;
		(*retoff) = newoff;
	}
out:
	if (reg)
		mutex_exit(&fp->f_rwlock);
	return (error);
}

off32_t
lseek32(int32_t fdes, off32_t off, int32_t stype)
{
	file_t *fp;
	int error;
	offset_t retoff;

	if ((fp = getf(fdes)) == NULL)
		return ((off32_t)set_errno(EBADF));

	/*
	 * lseek32 returns EOVERFLOW if we cannot represent the resulting
	 * offset from seek in a 32-bit off_t.
	 * The following routines are sensitive to sign extensions and
	 * calculations and if ever you change this make sure it works for
	 * special files.
	 *
	 * When VREG is not set we do the check for stype != SEEK_SET
	 * to send the unsigned value to lseek_common and not the sign
	 * extended value. (The maximum representable value is not
	 * checked by lseek_common for special files.)
	 */
	if (fp->f_vnode->v_type == VREG || stype != SEEK_SET)
		error = lseek32_common(fp, stype, (offset_t)off,
		    (offset_t)MAXOFF32_T, &retoff);
	else	/* stype == SEEK_SET */
		error = lseek32_common(fp, stype, (offset_t)(uint_t)off,
		    (offset_t)(uint_t)UINT_MAX, &retoff);

	releasef(fdes);
	if (!error)
		return ((off32_t)retoff);
	return ((off32_t)set_errno(error));
}

/*
 * 64-bit seeks from 32-bit applications
 */
offset_t
llseek32(int32_t fdes, uint32_t off1, uint32_t off2, int stype)
{
	file_t *fp;
	int error;
	offset_t retoff;
#if defined(_LITTLE_ENDIAN)
	offset_t off = ((u_offset_t)off2 << 32) | (u_offset_t)off1;
#else
	offset_t off = ((u_offset_t)off1 << 32) | (u_offset_t)off2;
#endif

	if ((fp = getf(fdes)) == NULL)
		error = EBADF;
	else {
		error = lseek32_common(fp, stype, off, MAXOFFSET_T, &retoff);
		releasef(fdes);
	}

	return (error? (offset_t)set_errno(error) : retoff);
}
#endif	/* _SYSCALL32_IMPL || _ILP32 */

#ifdef _LP64
/*
 * Seek on a file.
 *
 * Life is almost simple again (at least until we do 128-bit files ;-)
 * This is both 'lseek' and 'llseek' to a 64-bit application.
 */
off_t
lseek64(int fdes, off_t off, int stype)
{
	file_t *fp;
	vnode_t *vp;
	struct vattr vattr;
	int error;
	off_t old_off;
	offset_t new_off;
	int reg;

	if ((fp = getf(fdes)) == NULL)
		return ((off_t)set_errno(EBADF));

	vp = fp->f_vnode;
	reg = (vp->v_type == VREG);
	if (reg) {
		/*
		 * This is a POSIX requirement for atomicity
		 * of calls to read[v](), write[v]() and lseek()
		 * when operating on regular files.
		 */
		mutex_enter(&fp->f_rwlock);
	}
	new_off = off;

	switch (stype) {
	case SEEK_CUR:
		new_off += fp->f_offset;
		if (reg && off == 0) {	/* regular file, query only */
			error = 0;
			goto out;
		}
		break;

	case SEEK_END:
		vattr.va_mask = AT_SIZE;
		if ((error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL)) != 0)
			goto out;
		new_off += vattr.va_size;
		break;

	case SEEK_SET:
		break;

	case SEEK_DATA:
		/*
		 * Get and set the file pointer to the offset of the next
		 * data past "off"
		 */
		new_off = (offset_t)off;
		error = VOP_IOCTL(vp, _FIO_SEEK_DATA, (intptr_t)(&new_off),
		    FKIOCTL, kcred, NULL, NULL);
		if (error) {
			if (error != ENOTTY)
				goto out;
			/*
			 * The ioctl is not supported, check the supplied off
			 * is not past end of file
			 */
			vattr.va_mask = AT_SIZE;
			error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL);
			if (error)
				goto out;
			if (new_off >= (offset_t)vattr.va_size) {
				error = ENXIO;
				goto out;
			}
		}
		fp->f_offset = new_off;
		goto out;

	case SEEK_HOLE:
		/*
		 * Get and set the file pointer to the offset of the next
		 * hole past "off"
		 */
		new_off = off;
		error = VOP_IOCTL(vp, _FIO_SEEK_HOLE, (intptr_t)(&new_off),
		    FKIOCTL, kcred, NULL, NULL);
		if (error) {
			if (error != ENOTTY)
				goto out;
			/*
			 * ioctl is not supported, if the off is valid return
			 * the "virtual hole" at the end of the file.
			 */
			vattr.va_mask = AT_SIZE;
			error = VOP_GETATTR(vp, &vattr, 0, fp->f_cred, NULL);
			if (error)
				goto out;
			if (off < (offset_t)vattr.va_size) {
				new_off = (offset_t)vattr.va_size;
			} else {
				error = ENXIO;
				goto out;
			}
		}
		fp->f_offset = new_off;
		goto out;

	default:
		error = EINVAL;
		goto out;
	}

	old_off = fp->f_offset;
	if ((error = VOP_SEEK(vp, old_off, &new_off, NULL)) == 0) {
		/*
		 * lseek(fd, 0, SEEK_CUR) should not reset the file offset
		 * even if this is not a regular file.
		 */
		if (off != 0 || stype != SEEK_CUR)
			fp->f_offset = new_off;
	}
out:
	if (reg)
		mutex_exit(&fp->f_rwlock);
	releasef(fdes);
	return (error? (off_t)set_errno(error) : new_off);
}
#endif	/* _LP64 */
